packages/wrangler/src/config/validation-pages.ts

```diff
@@ -0,0 +1,126 @@
+/**
+ * Pages now supports configuration via `wrangler.toml`. As opposed to
+ * Workers however, Pages only supports a limited subset of all available
+ * configuration keys.
+ *
+ * This file contains all `wrangler.toml` validation things, specific to
+ * Pages.
+ */
+
+import { Diagnostics } from "./diagnostics";
+import { hasProperty } from "./validation-helpers";
+import type {
+	Config,
+	RawConfig,
+} from "./config";
+
+const supportedPagesConfigFields = [
+	"pages_build_output_dir",
+	"name",
+	"compatibility_date",
+	"compatibility_flags",
+	"placement",
+	"vars",
+	"kv_namespaces",
+	"queues", // `producers` ONLY
+	"r2_buckets",
+	"d1_databases",
+	"vectorize",
+	"services",
+	"analytics_engine_datasets",
+	"ai",
+	"ip",
+	"port",
+	"inspector_port",
+	"local_protocol",
+	"upstream_protocol",
+	"host"
+] as const;
+
+export function validatePagesConfig(
+	rawConfig: RawConfig,
+	config: Config
+): Diagnostics {
+	const diagnostics = new Diagnostics(
+		`Configuration file belonging to ⚡️ Pages ⚡️ project detected.
+		Running Pages configuration validation:`
+	);
+
+	validateMainField(config, diagnostics);
+	validatePagesEnvironmentNames(rawConfig, diagnostics);
+	validateUnsupportedFields(config, diagnostics);
+
+	return diagnostics;
+}
+
+/**
+ * Validate that configuration file doesn't declare "main", if
+ * "pages_build_output_dir" is present
+ */
+function validateMainField(config: Config, diagnostics: Diagnostics) {
+	if (hasProperty(config, "main")) {
+		diagnostics.errors.push(
+			`Configuration file cannot contain both both "main" and "pages_build_output_dir" configuration keys.
+			Please use "main" if you are deploying a Worker, or "pages_build_output_dir" if you are deploying a Pages project.`
+		);
+	}
+}
+
+/**
+ * Validate that no named-environments other than "preview" and "production"
+ * were declared in the configuration file for Pages
+ */
+function validatePagesEnvironmentNames(
```

Comment:
Pete suggested this validation should be performed inside `normalizeAndValidateConfig()`, which would remove the need to pass in `rawConfig` here. I tend to agree, so will address that

Reply:
I'm not sure about that—this feels like the right place for it given it's doing Pages specific validation?

Reply:
If it is Pages specific I agree with you @penalosa.

But doing it this way has a different downside, which is that this code needs to understand how RawConfig works and parse out the `env` property from there.

My suggestion was that it would be nice if these functions only cared about the processed `Config`. As such we could augment `normalizeAndValidateConfig()` with an extra parameter called `allowedEnvironments` (or something like that) which is not Pages specific per se but means that we don't need to pass the `rawConfig` into this function.

Moreover, this could feature of `normalizeAndValidateConfig()` could be extended to handle the environment fallback logic that we need for Pages - e.g. something like:

```
const allowedEnvs = [
 { name: "production", fallbackToTopLevel: true },
 { name: "preview", fallbackToTopLevel: true },
]
normalizeAndValidateConfig(rawConfig, configPath, args, allowedEnvs);
```

That kind of thing... open to other ideas but it seems awkward for us to have to have special logic that messes around with `RawConfig.env` outside of that function.

Reply:
removed the dependency on `RawConfig` for now

Reply:
actually...I want to come back this. so as I am looking at implementing env inheritance for Pages, I better understand now where you're coming from @petebacondarwin. told you my brain works in baby avocado increments 😅 🥑 

so assuming we would change the `normalizeAndValidateConfig()` fn signature to accept that extra `allowedEnvs` param, are you saying that the env name validation for Pages should take place inside that fn as well? If we don't, we basically risk creating that normalized config for an invalid Pages env name, only to then fail in `validatePagesConfig()`, which feels wasteful to me (tho IMHO we are already potentially doing wasteful config processing as we validate for Pages *after* we normalize, so maybe not a biggie). However, if we do, we leak Pages-specific validation things into `normalizeAndValidateConfig`, which is something I was hoping to prevent.

It seems to me, that either way we go abt it, there's no one way forward that feels 100% right and clean, at least to me not sure how y'all feel abt it, and there's some compromise that needs to happen between:
- keeping all Pages-specific validation code in one place
- avoid polluting `normalizeAndValidateConfig` with Pages things
- validating Pages against a normalized config

Reply:
I think my thinking around this was that it's unavoidable that we make some changes to `normalizeAndValidateConfig()`, but by splitting out the validation into Pages specific and workers specific we can minimise the changes to `normalizeAndValidateConfig()` to just the one line change outlined in the spec

Reply:
are you referring to the one liner env inheritance change here?

Reply:
@CarmenPopoviciu yes

Reply:
will leave this conversation unresolved for posterity, but AFAICT, we seem to be OK with the current implementation, so leaving as is. Appreciate the very constructive conversation around this <3